### PR TITLE
add docs: modularization intro & implement a Actuator

### PR DESCRIPTION
**Pull request by guoquanwu**
**What does this PR do?**

add docs: modularization intro & implement a Actuator

### CODE REVIEW

File: docs/Implement a customized Actuator.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+## 定义并注册合约
+
+目前 java-tron 支持的合约定义在 protocol 模块的 src/main/protos/core/contract 目录中，在这个目录下新建一个 math_contract.proto 文件并声明 SumContract，SumContract 的逻辑是将两个数值相加求和：
+```
+message SumContract {
+    int64 param1 = 1;
+    int64 param2 = 2;
+    bytes owner_address = 3;
+}
+```
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: use code highlight:

```protobuf
message SumContract {
    int64 param1 = 1;
    int64 param2 = 2;
    bytes owner_address = 3;
}
```

    ```protobuf
    message SumContract {
        int64 param1 = 1;
        int64 param2 = 2;
        bytes owner_address = 3;
    }
    ```

File: docs/Implement a customized Actuator.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+同时将新的合约类型注册在 src/main/protos/core/Tron.proto 文件的 Transaction 对象中，交易、账号、区块等重要的数据结构都定义在 Tron.proto 文件中：
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: `Transaction.Contract.ContractType`  枚举中

File: docs/Implement a customized Actuator.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    enum ContractType {
+      AccountCreateContract = 0;
+      TransferContract = 1;
+      ........
+      SumContract = 52;  
+    }
+  ...
+}
+```
+然后还需要注册一个方法来保证 rpc 能够接收并识别该类型合约的请求，目前 rpc 协议统一定义在 src/main/protos/api/api.proto，在 api.proto 中的 Wallet Service 新增 InvokeSum 接口：
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/rpc/gRPC/

File: docs/Implement a customized Actuator.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    };
+  };
+  ...
+};
+```
+最后重新编译修改过 proto 文件，可自行编译也可直接通过编译 java-tron 项目来编译 proto 文件：
+
+*目前 java-tron 采用的 protoc v3.4.0，自行编译时确保 protoc 版本一致。*
+```
+// recommended
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: Do not use `//` for command line comment.
use `#`

File: docs/Implement a customized Actuator.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+./gradlew build -x test
+
+// or build by yourself
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/by yourself/via protoc/

File: docs/Implement a customized Actuator.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        get: "/wallet/invokesum"
+      }
+    };
+  };
+  ...
+};
+```
+最后重新编译修改过 proto 文件，可自行编译也可直接通过编译 java-tron 项目来编译 proto 文件：
+
+*目前 java-tron 采用的 protoc v3.4.0，自行编译时确保 protoc 版本一致。*
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/采用的/采用的是/

File: docs/Implement a customized Actuator.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+protoc -I=src/main/protos -I=src/main/protos/core --java_out=src/main/java  Tron.proto
+protoc -I=src/main/protos/core/contract --java_out=src/main/java  math_contract.proto
+protoc -I=src/main/protos/api -I=src/main/protos/core -I=src/main/protos  --java_out=src/main/java api.proto
+```
+编译之后会更新 java_out 目录中对应的 java 文件。
+
+## 实现 SumActuator
+
+下面是基于编译更新后的 java 文件的 SumActuator 实现：
+```
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: code highlighting

File: docs/Implement a customized Actuator.md [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    }
+    return true;
+  }
+
+  /**
+   * define the rule to validate the contract
+   *
+   * this demo first check whether contract is null, then check whether ${any} is a instanceof SumContract,
+   * then validate the ownerAddress, finally check params are not less than 0.
+   */
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/check/checks/g
s/validate/validates/

> finally check whether the params are less than 0.

File: docs/Implement a customized Actuator.md [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+  @Override
+  public boolean validate() throws ContractValidateException {
+    // do some check
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: rm

File: docs/Implement a customized Actuator.md [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+      logger.debug(e.getMessage(), e);
+      throw new ContractValidateException(e.getMessage());
+    }
+    byte[] ownerAddress = sumContract.getOwnerAddress().toByteArray();
+    if (!DecodeUtil.addressValid(ownerAddress)) {
+      throw new ContractValidateException("Invalid ownerAddress!");
+    }
+    long param1 = sumContract.getParam1();
+    long param2 = sumContract.getParam2();
+    // do some check
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: rm

File: docs/Implement a customized Actuator.md [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+    if(param1 < 0 || param2 < 0){
+      logger.debug("negative number is not supported");
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * this methid return the ownerAddress
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/return/returns/

File: docs/Implement a customized Actuator.md [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+   * @return
+   * @throws InvalidProtocolBufferException
+   */
+  @Override
+  public ByteString getOwnerAddress() throws InvalidProtocolBufferException {
+    return any.unpack(SumContract.class).getOwnerAddress();
+  }
+
+  /**
+   * burn fee for a contract can reduce attacks like DDoS,
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/burn/burning/

Comment (andelf) [reply docs/Implement a customized Actuator.md]: s/,/./

File: docs/Implement a customized Actuator.md [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+   * choose the best strategy according to the business
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: > choose the best strategy according to business logic.

File: docs/Implement a customized Actuator.md [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+   * here return a contant just for demo
+   * @return
+   */
+  @Override
+  public long calcFee() {
+    return TRANSFER_FEE;
+  }
+}
+```
+为了简单起见 SumActuator 的执行结果直接输出至 log 文件中，对于有存储需求的情况，可以考虑创建一个新的 chainbase 来存储相应的数据。*具体如何创建 chainbase 不是本文讨论内容，相关文章后续发布，敬请期待*
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: *italic* style is seldom used in Chineses environment.

    *具体如何创建 chainbase 不是本文讨论内容，相关文章后续发布，敬请期待* =>
    (创建 chainbase 的方法将在后续相关文章中发布)

File: docs/Implement a customized Actuator.md [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    Args.clearParam();
+    appTest.shutdownServices();
+    appTest.shutdown();
+    context.destroy();
+    FileUtil.deleteDir(new File(dbPath));
+  }
+
+  @Test
+  public void sumActuatorTest() {
+    // this key define in config-localtest.conf as accountName=Sun
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: s/define/is defined/

File: docs/Implement a customized Actuator.md [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    // build contract
+    MathContract.SumContract.Builder builder = MathContract.SumContract.newBuilder();
+    builder.setParam1(1);
+    builder.setParam2(2);
+    builder.setOwnerAddress(ByteString.copyFrom(address));
+    MathContract.SumContract contract = builder.build();
+
+    // send contract and return trx
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: `trx` misleading. use `Transaction`

File: docs/Implement a customized Actuator.md [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+    if (ecKey == null || ecKey.getPrivKey() == null) {
+      logger.warn("Warning: Can't sign,there is no private key !!");
+      return null;
+    }
+    transaction = TransactionUtils.setTimestamp(transaction);
+    return TransactionUtils.sign(transaction, ecKey);
+  }
+}
+```
+运行 SumActuatorTest 测试类即可在log文件中看到 `SumActuator: param1 = 1, param2 = 2, sum = 3` 类似的输出字样，为了方便起见，本 demo 直接在 IDE 中执行，得到如下输出：
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: ~~为了方便起见，本 demo 直接在 IDE 中执行~~

then remove the timestamps of your logs.

File: docs/Implement a customized Actuator.md [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+18:40:16.461 INFO [actuator] 
+
+-------------------------------------------------
+|
+| SumActuator: param1 = 1, param2 = 2, sum = 3
+|
+-------------------------------------------------
+```
+
+至此，SumActuator 已基本实现完毕，这只是一个最简单的例子，真正的业务场景还需要做一些额外的工作，比如在 Wallet-Cli 中提供对应的合约支持、定制合适的 chainbase 存储等。
```
Comment (andelf) [inline docs/Implement a customized Actuator.md]: `Wallet-cli` for the repo name, use `wallet-cli`.

**Comment by andelf** [comment]
Code blocks, normal paragraphs, list items, headings should be joined with 1 empty line.

**Comment by guoquanwu** [comment]
all fixed, please review again, thanks

**Pull request merged**